core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,299 @@
+      txn.commitTransaction();
+      return null;
+    };
+
+    tasks.add(write1);
+    tasks.add(write2);
+    tasks.add(write3);
+    tasks.add(write4);
+    threadPool.invokeAll(tasks);
+    threadPool.shutdown();
+
+    Assert.assertEquals(4, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testRollBack() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    long snapshotId = table.currentSnapshot().snapshotId();
+    table.newFastAppend().appendFile(FILE_B).commit();
+
+    Assert.assertEquals(2, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    table.manageSnapshots().rollbackTo(snapshotId).commit();
+
+    Assert.assertEquals(1, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testMultipleTxnOperations() {
+    Snapshot snapshot;
+    Transaction txn = table.newTransaction();
+    txn.newOverwrite().addFile(FILE_A).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(1, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    Set<DataFile> toAddFiles = new HashSet<>();
+    Set<DataFile> toDeleteFiles = new HashSet<>();
+    toAddFiles.add(FILE_B);
+    toDeleteFiles.add(FILE_A);
+    txn.newRewrite().rewriteFiles(toDeleteFiles, toAddFiles).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(2, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.newReplacePartitions().addFile(FILE_C).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(3, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.newDelete().deleteFile(FILE_C).commit();
+    txn.commitTransaction();
+
+    Assert.assertEquals(4, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.newAppend().appendFile(FILE_C).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(5, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    snapshot = table.currentSnapshot();
+    txn = table.newTransaction();
+    txn.newOverwrite().addFile(FILE_D).deleteFile(FILE_C).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(6, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+
+    txn = table.newTransaction();
+    txn.expireSnapshots().expireOlderThan(snapshot.timestampMillis()).commit();
+    txn.commitTransaction();
+    Assert.assertEquals(6, TestTables.load(tableDir, "test").currentSnapshot().sequenceNumber());
+  }
+
+  @Test
+  public void testSequenceNumberForCherryPicking() {
+    table.newAppend()
+        .appendFile(FILE_A)
+        .commit();
+
+    // WAP commit
+    table.newAppend()
+        .appendFile(FILE_B)
+        .set("wap.id", "123456789")
```

Comment:
What is the purpose of a WAP ID? The commit is staged, so there is no reason to mix in WAP logic.

Reply:
Removed.